File: java/src/main/java/ml/wonwoo/springdatashowcase/product/ProductInitializerData.java
package ml.wonwoo.springdatashowcase.product;

import ml.wonwoo.springdatashowcase.data.InitializerData;
import ml.wonwoo.springdatashowcase.product.Book.BookType;
import ml.wonwoo.springdatashowcase.product.Disc.DiscType;
import org.javamoney.moneta.Money;
import org.springframework.stereotype.Component;

import javax.money.MonetaryAmount;

import static ml.wonwoo.springdatashowcase.support.Currencies.USD;

@Component
public class ProductInitializerData implements InitializerData {

    private final BookCatalog bookCatalog;
    private final DiscCatalog discCatalog;

    public ProductInitializerData(BookCatalog bookCatalog, DiscCatalog discCatalog) {
        this.bookCatalog = bookCatalog;
        this.discCatalog = discCatalog;
    }

    @Override
    public void initialize() {

        bookCatalog.save(new Book("Effective Java", "http://localhost:8080",
            "Joshua Bloch", Money.of(3.99, USD), BookType.SOFTWARE));

        bookCatalog.save(new Book("Effective Java", "http://localhost:8080",
            "Joshua Bloch", Money.of(8.99, USD), BookType.SOFTWARE));

        bookCatalog.save(new Book("Back to the Future", "http://localhost:8080",
            "Michael Klastorin", Money.of(6.99, USD), BookType.FANTASY));

        discCatalog.save(new Disc("Back to the Future", "http://localhost:8080",
            "Sci-Fi", Money.of(5.99, USD), DiscType.BLURAY));

        Books authors = bookCatalog.findByAuthor("Joshua Bloch");
        MonetaryAmount total = authors.getTotal();
        System.out.println(total);

    }
}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/product/ProductInitializerData.kt
package ml.wonwoo.springdatashowcase.product

import ml.wonwoo.springdatashowcase.data.InitializerData
import ml.wonwoo.springdatashowcase.support.USD
import ml.wonwoo.springdatashowcase.support.getTotal
import ml.wonwoo.springdatashowcase.support.logger
import org.javamoney.moneta.Money
import org.springframework.stereotype.Component

@Component
class ProductInitializerData(private val bookCatalog: BookCatalog,
                             private val discCatalog: DiscCatalog) : InitializerData {

    val logger = logger<ProductInitializerData>()

    override fun initialize() {

        bookCatalog.save(Book("Effective Java", Money.of(3.99, USD),
            "http://localhost:8080", "Joshua Bloch", BookType.SOFTWARE))

        bookCatalog.save(Book("Effective Java", Money.of(8.99, USD),
            "http://localhost:8080", "Joshua Bloch", BookType.SOFTWARE))

        bookCatalog.save(Book("Back to the Future", Money.of(6.99, USD),
            "http://localhost:8080", "Michael Klastorin", BookType.FANTASY))

        discCatalog.save(Disc("Back to the Future", Money.of(5.99, USD),
            "http://localhost:8080", "Sci-Fi", DiscType.BLURAY))

        val books = bookCatalog.findByAuthor("Joshua Bloch")

        logger.info("${books.getTotal()}")

        val discs = discCatalog.findByGenre("Sci-Fi")

        logger.info("${discs.getTotal()}")

        val bookNames = bookCatalog.findName("Effective Java")

        logger.info("$bookNames")

        val bookNames2 = bookCatalog.findByName("Effective Java")

        logger.info("$bookNames2")
    }

}
